File: 15-3sum/15-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */


// sort array, do two sum II for each element
const threeSum = function(nums) {
    const result = [];
    nums.sort((a,b) => a - b);
    for (let i = 0; i < nums.length; i++) {
        // eliminate duplicates by skipping duplicates (i-1)
        if (i > 0 && nums[i] === nums[i-1]) {
            continue;
        }
        // everything <= i has been evaluated
        let left = i + 1,
            right = nums.length - 1;
        // perform sorted two sum
        while (left < right) {
            const curSum = nums[i] + nums[left] + nums[right];
            if (curSum < 0) left++;
            else if (curSum > 0) right--;
            else {
                result.push([nums[i], nums[left], nums[right]]);
                // only need to update one pointer
                // skip if same value to avoid duplicates
                // right pointer auto shift from logic above
                left++;
                while (nums[left] === nums[left-1] && left < right) {
                    left++;
                }
            }
        }
    }
    return result;
};

File: medium/15-3sum/15-3sum.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
    nums.sort!
    res = []
    nums.each_with_index do |n, i|
        next if i > 0 && n == nums[i - 1]
        l = i + 1
        r = nums.size - 1
        while l < r do 
            cur_sum = n + nums[l] + nums[r]
            if cur_sum < 0
                l += 1
            elsif cur_sum > 0
                r -= 1 
            else
                res.push([n, nums[l], nums[r]])
                l += 1
                while nums[l] == nums[l - 1] && l < r do 
                    l += 1 
                end
            end
        end
    end
     p res
end